source/common/http/filter_manager.cc

```diff
@@ -183,11 +183,12 @@ bool ActiveStreamFilterBase::commonHandleAfterDataCallback(FilterDataStatus stat
       buffer_was_streaming = status == FilterDataStatus::StopIterationAndWatermark;
       commonHandleBufferData(provided_data);
     } else if (complete() && !hasTrailers() && !bufferedData()) {
-      // If this filter is doing StopIterationNoBuffer and this stream is terminated with a zero
```

Comment:
Ah, now I have the context for this change.
ccing @snowp for a second opinion but in mine, if we end up with a sendLocalReply in call stack of the filter chain, we shouldn't be calling commonHandleAfterDataCallback at all, we should short circuit and early return from FilterManager::decodeData

I'd be inclined to revert that PR and add that short circuit here, but again let's see what @snowp or @mattklein123 think.

Reply:
Sorry I don't think I have enough context. What PR are we reverting?

Reply:
https://github.com/envoyproxy/envoy/pull/15075

Basically sendLocalReply is supposed to Just Work (TM) in all cases. If a filter calls sendLocalReply after the response filter chain has started, it should result in either the local reply being sent straight to the codec (as we can't traverse the response filter chain again safely) or stream reset (if the headers have been shipped downstream already)

Here in this PR we try to sendLocalReply under the stack of encodeData, and it turns it it doesn't work becasue the commonHandleAfterDataCallback assertions are failing.  #15075 was an attempt to fix the asserts to not break, but IMO we shouldn't call commonHandleAfterDataCallback if we've aborted the original reply (in encodedata) in favor of sendLocalReply being shipped immediately out.

Reply:
OK yeah +1 on reverting the other PR and then let's just sort it all out in this PR? I think that would be easier to review agreed.

Reply:
I'd be inclined to track when end_stream=true is sent to the codec (probably in FilterManager::maybeEndEncode) and check against that?  Seems more general purpose.

Reply:
Added a new state variable to short-circuit the logic.

I tried finding a more general approach (by using `end_stream_`), but there are cases where end_stream=false but the data buffer still needs to be handled.

I do not have enough understanding of the FilterManager code to feel confident about making larger changes to it. If the current non-general approach works, can we go move forward with it?